Micronaut Data Kapt compilation flakiness

### Expected Behavior

Gradle task kaptKotlin pass

### Actual Behaviour

Gradle task kaptKotlin fails from time to time with the following error:
```kotlin
> Task :kaptKotlin
error: Unable to implement Repository method: ConnectionRepository.save(Object entity). No possible implementations found.
```

### Steps To Reproduce

Don't have the stable reproduce, it's flaky.
It looks like a classpath or annotation processing order issue.
```kotlin
package org.myorg.dao

import org.myorg.dto.ConnectionId
import org.myorg.model.Connection
import io.micronaut.data.mongodb.annotation.MongoRepository
import io.micronaut.data.repository.CrudRepository

@MongoRepository
interface ConnectionRepository : CrudRepository<Connection, ConnectionId>
```
```kotlin
@MappedEntity
data class Connection(
    @field:Id
    @GeneratedValue
    var id: ConnectionId = "",
    val name: String
)
```
```kotlin
typealias ConnectionId = String
```

```kotlin
plugins {
    id("org.jetbrains.kotlin.jvm") version "1.6.10"
    id("org.jetbrains.kotlin.kapt") version "1.6.10"
    id("org.jetbrains.kotlin.plugin.allopen") version "1.6.10"
    id("io.micronaut.application") version "3.3.0"
    id("io.gitlab.arturbosch.detekt") version "1.19.0"
}
...
micronaut {
    runtime("netty")
    testRuntime("junit5")
    processing {
        incremental(true)
        annotations("org.myorg.*")
    }
}
...
dependencies {
    kapt("io.micronaut:micronaut-http-validation")
    kapt("io.micronaut.data:micronaut-data-document-processor:3.3.0")
    kapt("io.micronaut.openapi:micronaut-openapi")
    implementation("io.micronaut:micronaut-http-client")
    implementation("io.micronaut:micronaut-jackson-databind")
    implementation("io.micronaut:micronaut-runtime")
    implementation("io.micronaut.kotlin:micronaut-kotlin-runtime")
    implementation("io.micronaut.data:micronaut-data-mongodb:3.3.0")
    implementation("io.micronaut.mongodb:micronaut-mongo-sync")
    implementation("io.swagger.core.v3:swagger-annotations")
    implementation("jakarta.annotation:jakarta.annotation-api")
    implementation("org.jetbrains.kotlin:kotlin-reflect:$kotlinVersion")
    implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlinVersion")
    runtimeOnly("ch.qos.logback:logback-classic")
    implementation("io.micronaut:micronaut-validation")
    implementation("io.projectreactor:reactor-core:3.4.14")

    runtimeOnly("com.fasterxml.jackson.module:jackson-module-kotlin")

    detektPlugins("io.gitlab.arturbosch.detekt:detekt-formatting:1.19.0")

    testImplementation("com.tngtech.archunit:archunit-junit5:0.23.1")
}
```

### Environment Information

- OS Ubuntu 20.04
- Java 17

### Example Application

_No response_

### Version

3.3.4

We got (I think) the same error in our build. The interesting part is that it happened with Maven. We run maven with `-T 3`.

Today was the first occurence after ~20 days, with multiple daily builds. So it looks quite rare. Possibly related to the annotationProcessor?

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.2:compile (default-compile) on project xxxxxxxxx: Compilation failure
[ERROR] Unable to implement Repository method: XxxRepository.save(Object entity). No possible implementations found.
```

We use micronaut-data with r2dbc. Version 3.2.2

Collected more logs
Full log in the attached file: 
[kaptKotlin.log](https://github.com/micronaut-projects/micronaut-data/files/8358141/kaptKotlin.log)

```
20:19:44  2022-03-27T17:19:44.471+0000 [ERROR] [system.err] error: Unable to implement Repository method: ConnectionRepository.save(Object entity). No possible implementations found.
20:19:44  2022-03-27T17:19:44.472+0000 [ERROR] [system.err] Note: Generating OpenAPI Documentation
20:19:44  2022-03-27T17:19:44.472+0000 [ERROR] [system.err] Note: Writing OpenAPI YAML to destination: /home/.../build/tmp/kapt3/classes/main/META-INF/swagger/connection-manager-service-0.0.3.yml
20:19:44  2022-03-27T17:19:44.472+0000 [ERROR] [system.err] Note: Writing OpenAPI View to destination: /home/.../build/tmp/kapt3/classes/main/META-INF/swagger/views/redoc/index.html
20:19:44  2022-03-27T17:19:44.473+0000 [ERROR] [system.err] Note: Writing OpenAPI View to destination: /home/.../build/tmp/kapt3/classes/main/META-INF/swagger/views/rapidoc/index.html
20:19:44  2022-03-27T17:19:44.473+0000 [ERROR] [system.err] Note: Writing OpenAPI View to destination: /home/.../build/tmp/kapt3/classes/main/META-INF/swagger/views/swagger-ui/index.html
20:19:44  2022-03-27T17:19:44.473+0000 [ERROR] [system.err] Note: Creating bean classes for 24 type elements
...
20:19:44  2022-03-27T17:19:44.599+0000 [DEBUG] [org.jetbrains.kotlin.gradle.plugin.KotlinGradleBuildServices$Inject] [KOTLIN] Disposed KotlinGradleBuildServices
20:19:44  2022-03-27T17:19:44.610+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 
20:19:44  2022-03-27T17:19:44.610+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] FAILURE: Build failed with an exception.
20:19:44  2022-03-27T17:19:44.610+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 
20:19:44  2022-03-27T17:19:44.610+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * What went wrong:
20:19:44  2022-03-27T17:19:44.610+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] Execution failed for task ':kaptKotlin'.
20:19:44  2022-03-27T17:19:44.610+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] > A failure occurred while executing org.jetbrains.kotlin.gradle.internal.KaptWithoutKotlincTask$KaptExecutionWorkAction
20:19:44  2022-03-27T17:19:44.610+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter]    > java.lang.reflect.InvocationTargetException (no error message)
20:19:44  2022-03-27T17:19:44.610+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 
20:19:44  2022-03-27T17:19:44.611+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * Try:
20:19:44  2022-03-27T17:19:44.611+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] > Run with --scan to get full insights.
20:19:44  2022-03-27T17:19:44.611+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 
20:19:44  2022-03-27T17:19:44.611+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] * Exception is:
20:19:44  2022-03-27T17:19:44.611+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':kaptKotlin'.
20:19:44  2022-03-27T17:19:44.611+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:145)
20:19:44  2022-03-27T17:19:44.611+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:282)
20:19:44  2022-03-27T17:19:44.611+0000 [ERROR] [org.gradle.internal.buildevents.BuildExceptionReporter] 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:143)
```

--- a/config/accepted-api-changes.json
+++ b/config/accepted-api-changes.json
@@ -58,5 +58,25 @@
         "type": "io.micronaut.transaction.jdbc.DataSourceTransactionManager",
         "member": "Class io.micronaut.transaction.jdbc.DataSourceTransactionManager",
         "reason": "Internal method"
+    },
+    {
+    "type": "io.micronaut.data.processor.visitors.finders.AbstractPrefixPatternMethodMatcher",
+    "member": "Class io.micronaut.data.processor.visitors.finders.AbstractPrefixPatternMethodMatcher",
+    "reason": "Class is considered internal API"
+},
+    {
+        "type": "io.micronaut.data.processor.visitors.finders.AbstractPrefixPatternMethodMatcher",
+        "member": "Constructor io.micronaut.data.processor.visitors.finders.AbstractPrefixPatternMethodMatcher(java.lang.String[])",
+        "reason": "Class is considered internal API"
+    },
+    {
+        "type": "io.micronaut.data.processor.visitors.finders.SaveEntityMethodMatcher",
+        "member": "Class io.micronaut.data.processor.visitors.finders.SaveEntityMethodMatcher",
+        "reason": "Class is considered internal API"
+    },
+    {
+        "type": "io.micronaut.data.processor.visitors.finders.SaveEntityMethodMatcher",
+        "member": "Field PREFIXES",
+        "reason": "Class is considered internal API"
     }
 ]

--- a/data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/AbstractPrefixPatternMethodMatcher.java
+++ b/data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/AbstractPrefixPatternMethodMatcher.java
@@ -17,12 +17,12 @@ package io.micronaut.data.processor.visitors.finders;
 
 import io.micronaut.core.annotation.Experimental;
 import io.micronaut.core.annotation.NonNull;
-import io.micronaut.core.util.ArrayUtils;
 import io.micronaut.data.processor.visitors.MethodMatchContext;
 
-import java.util.Arrays;
 import java.util.Comparator;
+import java.util.List;
 import java.util.regex.Pattern;
+import java.util.stream.Collectors;
 
 /**
  * The matcher based on a simple patter.
@@ -33,13 +33,14 @@ import java.util.regex.Pattern;
 @Experimental
 public abstract class AbstractPrefixPatternMethodMatcher implements MethodMatcher {
 
+    private static final Comparator<String> PREFIX_COMPARATOR = Comparator.comparingInt(String::length).thenComparing(String::compareTo).reversed();
+
     protected final Pattern pattern;
 
-    protected AbstractPrefixPatternMethodMatcher(@NonNull String... prefixes) {
-        if (ArrayUtils.isEmpty(prefixes)) {
+    protected AbstractPrefixPatternMethodMatcher(@NonNull List<String> prefixes) {
+        if (prefixes.isEmpty()) {
             throw new IllegalArgumentException("At least one prefix required");
         }
-        Arrays.sort(prefixes, Comparator.comparingInt(String::length).thenComparing(String::compareTo).reversed());
         this.pattern = computePattern(prefixes);
     }
 
@@ -57,13 +58,15 @@ public abstract class AbstractPrefixPatternMethodMatcher implements MethodMatche
      * Handle the match.
      *
      * @param matchContext The match context
-     * @param matcher      The matcher
+     * @param matcher The matcher
      * @return The method matcher
      */
     protected abstract MethodMatch match(MethodMatchContext matchContext, java.util.regex.Matcher matcher);
 
-    private static Pattern computePattern(String[] prefixes) {
-        String prefixPattern = String.join("|", prefixes);
+    private static Pattern computePattern(List<String> prefixes) {
+        String prefixPattern = prefixes.stream()
+                .sorted(PREFIX_COMPARATOR)
+                .collect(Collectors.joining("|"));
         return Pattern.compile("^((" + prefixPattern + ")(\\S*?))$");
     }
 

--- a/data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/AbstractSpecificationMethodMatcher.java
+++ b/data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/AbstractSpecificationMethodMatcher.java
@@ -24,6 +24,8 @@ import io.micronaut.inject.ast.MethodElement;
 import io.micronaut.inject.ast.ParameterElement;
 import io.micronaut.inject.visitor.VisitorContext;
 
+import java.util.Arrays;
+
 /**
  * Abstract superclass for specification methods.
  *
@@ -39,7 +41,7 @@ public abstract class AbstractSpecificationMethodMatcher extends AbstractPrefixP
      * @param prefixes The method prefixes to match
      */
     protected AbstractSpecificationMethodMatcher(String... prefixes) {
-        super(prefixes);
+        super(Arrays.asList(prefixes));
     }
 
     @Override

--- a/data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/SaveEntityMethodMatcher.java
+++ b/data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/SaveEntityMethodMatcher.java
@@ -27,6 +27,8 @@ import io.micronaut.inject.ast.ClassElement;
 import io.micronaut.inject.ast.ParameterElement;
 
 import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
 import java.util.Map;
 
 import static io.micronaut.data.processor.visitors.finders.FindersUtils.getInterceptorElement;
@@ -39,7 +41,9 @@ import static io.micronaut.data.processor.visitors.finders.FindersUtils.getInter
  */
 public class SaveEntityMethodMatcher extends AbstractPrefixPatternMethodMatcher {
 
-    public static final String[] PREFIXES = {"save", "persist", "store", "insert"};
+    public static final List<String> PREFIXES = Collections.unmodifiableList(
+            Arrays.asList("save", "persist", "store", "insert")
+    );
 
     /**
      * The default constructor.

